kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala

```diff
@@ -484,6 +484,178 @@ sealed abstract class Resource[F[_], +A] {
     new (F ~> F) {
       override def apply[B](gb: F[B]): F[B] = surround(gb)
     }
+
+  def forceR[B](that: Resource[F, B])(implicit F: MonadCancel[F, Throwable]): Resource[F, B] =
+    Resource applyFull { poll =>
+      poll(this.use_ !> that.allocated) map { p =>
+        // map exists on tuple in Scala 3 and has the wrong type
+        Functor[(B, *)].map(p)(fin => (_: Resource.ExitCase) => fin)
+      }
+    }
+
+  def onCancel[B >: A](fin: Resource[F, Unit])(
+      implicit F: MonadCancel[F, Throwable]): Resource[F, B] =
+    Resource applyFull { poll =>
+      poll(this.allocated).onCancel(fin.use_) map { p =>
+        Functor[(B, *)].map(p)(fin => (_: Resource.ExitCase) => fin)
+      }
+    }
+
+  def guaranteeCase[B >: A](fin: Outcome[Resource[F, *], Throwable, B] => Resource[F, Unit])(
+      implicit F: MonadCancel[F, Throwable]): Resource[F, B] =
+    Resource applyFull { poll =>
+      val back = poll(this.allocated) guaranteeCase {
+        case Outcome.Succeeded(ft) =>
+          fin(Outcome.Succeeded(Resource.eval(ft.map(_._1)))).use_ handleErrorWith { e =>
+            ft.flatMap(_._2).handleError(_ => ()) >> F.raiseError(e)
+          }
+
+        case Outcome.Errored(e) =>
+          fin(Outcome.Errored(e)).use_.handleError(_ => ())
+
+        case Outcome.Canceled() =>
+          fin(Outcome.Canceled()).use_
+      }
+
+      back.map(p => Functor[(B, *)].map(p)(fu => (_: Resource.ExitCase) => fu))
+    }
+
+  /*
+   * 1. If the scope containing the *fiber* terminates:
+   *   a) if the fiber is incomplete, run its inner finalizers when it completes
+   *   b) if the fiber is succeeded, run its finalizers
+   * 2. If the fiber is canceled or errored, finalize
+   * 3. If the fiber succeeds and .cancel won the race, finalize eagerly and
+   *    `join` results in `Canceled()`
+   * 4. If the fiber succeeds and .cancel lost the race or wasn't called,
+   *    finalize naturally when the containing scope ends, `join` returns
+   *    the value
+   */
+  def start[B >: A](
+      implicit F: Concurrent[F]): Resource[F, Fiber[Resource[F, *], Throwable, B]] = {
```

Comment:
```suggestion
  def start(
      implicit F: Concurrent[F]): Resource[F, Fiber[Resource[F, *], Throwable, A @uncheckedVariance]] = {
```

This one also needs `uncheckedVariance`. Basically this comes from the fact that `Fiber` and `Outcome` are both invariant, but their parameter is used in entirely covariant contexts when the type constructor is also covariant, meaning that `uncheckedVariance` is sound.